File: scripts/changelog-generation/rewrite-changelog-section.js
'use strict';
/**
 * rewrite-changelog-section.js
 *
 * Rewrites the target section in CHANGELOG.md.
 *
 * In 'unreleased' mode:
 *   - Replaces the content of the ## [Unreleased] section with the new content.
 *   - Preserves all other sections exactly.
 *
 * In 'release' mode:
 *   - Replaces (or inserts after ## [Unreleased]) the ## [x.y.z] - YYYY-MM-DD section.
 *   - Preserves the ## [Unreleased] section exactly (does not modify it).
 *   - Preserves all other sections exactly.
 *   - Preserves the link footer exactly.
 *
 * Usage (CLI):
 *   node scripts/changelog-generation/rewrite-changelog-section.js \
 *     --mode unreleased|release \
 *     [--target-version 0.14.4] \
 *     [--section-content "### Changes\n\n- Foo (#123)"] \
 *     [--section-file path/to/section.md] \
 *     [--changelog CHANGELOG.md]
 *
 * Env vars (alternative to CLI args):
 *   MODE, TARGET_VERSION, SECTION_CONTENT, SECTION_FILE, CHANGELOG_PATH
 *
 * Outputs JSON: { success: boolean, changelogPath: string, section: string }
 *
 * Exports: rewriteUnreleased, rewriteRelease, rewriteSection (for unit testing)
 */

const fs = require('node:fs');
const path = require('node:path');

// ---------------------------------------------------------------------------
// Core parsing helpers
// ---------------------------------------------------------------------------

/**
 * Splits CHANGELOG.md content into logical blocks:
 *   - 'header':  text before the first ## header (may be empty)
 *   - 'section': { header: string, body: string } — one per ## section
 *   - 'footer':  text after the last section (link definitions etc.)
 *
 * @param {string} content
 * @returns {{ preamble: string, sections: Array<{ header: string, body: string }>, footer: string }}
 */
function parseChangelog(content) {
  const lines = content.split('\n');
  const sections = [];
  let preamble = '';
  let footer = '';

  let currentHeader = null;
  let currentBodyLines = [];
  let inSections = false;

  for (const line of lines) {
    if (line.startsWith('## ')) {
      if (currentHeader !== null) {
        sections.push({ header: currentHeader, body: currentBodyLines.join('\n') });
      } else if (!inSections) {
        // Everything before the first ## section is preamble
        preamble = currentBodyLines.join('\n');
      }
      currentHeader = line;
      currentBodyLines = [];
      inSections = true;
    } else if (inSections) {
      currentBodyLines.push(line);
    } else {
      currentBodyLines.push(line);
    }
  }

  if (currentHeader !== null) {
    sections.push({ header: currentHeader, body: currentBodyLines.join('\n') });
  } else {
    // No sections found at all
    preamble = currentBodyLines.join('\n');
  }

  // The footer is the trailing block of link definitions after the last section.
  // Link definitions look like: [label]: url
  // We extract them from the last section's body tail if they appear after a blank line gap.
  if (sections.length > 0) {
    const lastSection = sections[sections.length - 1];
    const bodyLines = lastSection.body.split('\n');

    // Find where the link footer starts: a line matching [label]: https://...
    let footerStart = -1;
    for (let i = bodyLines.length - 1; i >= 0; i--) {
      if (/^\[.+\]:\s*https?:\/\//.test(bodyLines[i])) {
        footerStart = i;
      } else if (bodyLines[i].trim() !== '') {
        break;
      }
    }

    if (footerStart !== -1) {
      footer = bodyLines.slice(footerStart).join('\n');
      lastSection.body = bodyLines.slice(0, footerStart).join('\n');
    }
  }

  return { preamble, sections, footer };
}

/**
 * Serialises a parsed changelog back to a string.
 *
 * @param {{ preamble: string, sections: Array<{ header: string, body: string }>, footer: string }} parsed
 * @returns {string}
 */
function serialiseChangelog({ preamble, sections, footer }) {
  const parts = [];

  if (preamble && preamble.trim()) {
    parts.push(preamble);
  }

  for (const section of sections) {
    parts.push(section.header);
    if (section.body !== undefined) {
      parts.push(section.body);
    }
  }

  if (footer && footer.trim()) {
    parts.push(footer);
  }

  return parts.join('\n');
}

/**
 * Normalises new section content:
 *   - Ensures it starts with a blank line after the header (will be injected separately).
 *   - Trims trailing blank lines but preserves a single trailing newline.
 *
 * @param {string} content — the body content (without the ## header line)
 * @returns {string}
 */
function normaliseSectionBody(content) {
  // Ensure the body starts with a single newline (header will be joined with '\n' by serialiseChangelog)
  const trimmed = content.replace(/^\n+/, '').trimEnd();
  return '\n' + trimmed + '\n';
}

// ---------------------------------------------------------------------------
// Rewrite operations
// ---------------------------------------------------------------------------

/**
 * Rewrites the ## [Unreleased] section with the given body content.
 *
 * @param {string} changelogContent — current CHANGELOG.md content
 * @param {string} newBody — new content for the section body (without the header line)
 * @returns {string} — updated CHANGELOG.md content
 */
function rewriteUnreleased(changelogContent, newBody) {
  const parsed = parseChangelog(changelogContent);

  const idx = parsed.sections.findIndex((s) => s.header === '## [Unreleased]');
  if (idx === -1) {
    throw new Error('## [Unreleased] section not found in CHANGELOG.md');
  }

  parsed.sections[idx].body = normaliseSectionBody(newBody);

  return serialiseChangelog(parsed);
}

/**
 * Rewrites the ## [x.y.z] - YYYY-MM-DD section for the given version,
 * inserting it after ## [Unreleased] if it doesn't exist yet.
 * The ## [Unreleased] section is left unchanged.
 *
 * @param {string} changelogContent — current CHANGELOG.md content
 * @param {string} targetVersion — e.g. '0.14.4'
 * @param {string} releaseDate — e.g. '2026-04-16'
 * @param {string} newBody — new content for the release section body (without the header line)
 * @returns {string} — updated CHANGELOG.md content
 */
function rewriteRelease(changelogContent, targetVersion, releaseDate, newBody) {
  const parsed = parseChangelog(changelogContent);

  const releaseHeader = `## [${targetVersion}] - ${releaseDate}`;
  const headerPrefix = `## [${targetVersion}]`;

  // Find existing release section (match by version prefix, ignoring date)
  let releaseIdx = parsed.sections.findIndex((s) => s.header.startsWith(headerPrefix));

  if (releaseIdx === -1) {
    // Insert after ## [Unreleased]
    const unreleasedIdx = parsed.sections.findIndex((s) => s.header === '## [Unreleased]');
    const insertAfter = unreleasedIdx !== -1 ? unreleasedIdx : -1;
    const newSection = { header: releaseHeader, body: normaliseSectionBody(newBody) };
    if (insertAfter !== -1) {
      parsed.sections.splice(insertAfter + 1, 0, newSection);
    } else {
      parsed.sections.unshift(newSection);
    }
  } else {
    parsed.sections[releaseIdx].header = releaseHeader;
    parsed.sections[releaseIdx].body = normaliseSectionBody(newBody);
  }

  return serialiseChangelog(parsed);
}

/**
 * Unified entry point: rewrites the appropriate section based on mode.
 *
 * @param {object} params
 * @param {string} params.changelogPath   — path to CHANGELOG.md
 * @param {string} params.mode            — 'unreleased' | 'release'
 * @param {string} [params.targetVersion] — required when mode === 'release'
 * @param {string} [params.releaseDate]   — ISO date string; defaults to today
 * @param {string} params.newSectionBody  — the new body content (without header)
 * @returns {{ success: boolean, changelogPath: string, section: string }}
 */
function rewriteSection({ changelogPath, mode, targetVersion, releaseDate, newSectionBody }) {
  const absPath = path.resolve(changelogPath);
  const changelogContent = fs.readFileSync(absPath, 'utf8');

  const date = releaseDate || new Date().toISOString().split('T')[0];

  let updated;
  let sectionHeader;

  if (mode === 'release') {
    if (!targetVersion) {
      throw new Error('targetVersion is required when mode === "release"');
    }
    updated = rewriteRelease(changelogContent, targetVersion, date, newSectionBody);
    sectionHeader = `## [${targetVersion}] - ${date}`;
  } else {
    updated = rewriteUnreleased(changelogContent, newSectionBody);
    sectionHeader = '## [Unreleased]';
  }

  fs.writeFileSync(absPath, updated, 'utf8');

  return { success: true, changelogPath: absPath, section: sectionHeader };
}

// ---------------------------------------------------------------------------
// CLI entry point
// ---------------------------------------------------------------------------

if (require.main === module) {
  // Parse CLI args
  const args = process.argv.slice(2);
  const argMap = {};
  for (let i = 0; i < args.length; i += 2) {
    const key = args[i].replace(/^--/, '').replace(/-([a-z])/g, (_, c) => c.toUpperCase());
    argMap[key] = args[i + 1];
  }

  const mode = argMap.mode || process.env.MODE || 'unreleased';
  const targetVersion = argMap.targetVersion || process.env.TARGET_VERSION || '';
  const changelogPath = argMap.changelog || process.env.CHANGELOG_PATH || 'CHANGELOG.md';
  const sectionFile = argMap.sectionFile || process.env.SECTION_FILE || '';

  let newSectionBody = argMap.sectionContent || process.env.SECTION_CONTENT || '';

  if (!newSectionBody && sectionFile) {
    try {
      newSectionBody = fs.readFileSync(sectionFile, 'utf8');
    } catch (err) {
      const result = { success: false, changelogPath, section: '', error: `Cannot read section file: ${err.message}` };
      process.stdout.write(JSON.stringify(result, null, 2) + '\n');
      process.exit(1);
    }
  }

  if (!newSectionBody) {
    const result = { success: false, changelogPath, section: '', error: 'No section content provided (use --section-content or --section-file)' };
    process.stdout.write(JSON.stringify(result, null, 2) + '\n');
    process.exit(1);
  }

  try {
    const result = rewriteSection({
      changelogPath,
      mode,
      targetVersion,
      newSectionBody,
    });
    process.stdout.write(JSON.stringify(result, null, 2) + '\n');
    process.exit(0);
  } catch (err) {
    const result = { success: false, changelogPath, section: '', error: err.message };
    process.stdout.write(JSON.stringify(result, null, 2) + '\n');
    process.exit(1);
  }
}

module.exports = { parseChangelog, serialiseChangelog, rewriteUnreleased, rewriteRelease, rewriteSection };

File: scripts/changelog-generation/rewrite_changelog_section.go
// Licensed to Elasticsearch B.V. under one or more contributor
// license agreements. See the NOTICE file distributed with
// this work for additional information regarding copyright
// ownership. Elasticsearch B.V. licenses this file to you under
// the Apache License, Version 2.0 (the "License"); you may
// not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

package main

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"slices"
	"strings"
	"time"
)

// ---------------------------------------------------------------------------
// Core parsing helpers
// ---------------------------------------------------------------------------

// changelogSection represents a ## section in CHANGELOG.md.
type changelogSection struct {
	header string
	body   string
	// bodySet indicates that the body field was explicitly set (even if empty string).
	// This mirrors the JS behaviour where `body !== undefined`.
	bodySet bool
}

// parsedChangelog is the result of parsing CHANGELOG.md into logical blocks.
type parsedChangelog struct {
	preamble string
	sections []changelogSection
	footer   string
}

// headerUnreleased is the standard Unreleased section header in CHANGELOG.md.
const headerUnreleased = "## [Unreleased]"

// linkDefRE matches a link definition line: [label]: https://...
var linkDefRE = regexp.MustCompile(`^\[.+\]:\s*https?://`)

// parseChangelog splits CHANGELOG.md content into preamble, sections, and footer.
func parseChangelog(content string) parsedChangelog {
	lines := strings.Split(content, "\n")
	var sections []changelogSection
	preamble := ""

	var currentHeader string
	var currentBodyLines []string
	inSections := false
	hasHeader := false

	for _, line := range lines {
		if strings.HasPrefix(line, "## ") {
			if hasHeader {
				sections = append(sections, changelogSection{
					header:  currentHeader,
					body:    strings.Join(currentBodyLines, "\n"),
					bodySet: true,
				})
			} else if !inSections {
				// Everything before the first ## section is preamble
				preamble = strings.Join(currentBodyLines, "\n")
			}
			currentHeader = line
			currentBodyLines = nil
			inSections = true
			hasHeader = true
		} else {
			currentBodyLines = append(currentBodyLines, line)
		}
	}

	if hasHeader {
		sections = append(sections, changelogSection{
			header:  currentHeader,
			body:    strings.Join(currentBodyLines, "\n"),
			bodySet: true,
		})
	} else {
		// No sections found at all
		preamble = strings.Join(currentBodyLines, "\n")
	}

	// The footer is the trailing block of link definitions after the last section.
	// Link definitions look like: [label]: https://...
	// We extract them from the last section's body tail.
	footer := ""
	if len(sections) > 0 {
		lastSection := &sections[len(sections)-1]
		bodyLines := strings.Split(lastSection.body, "\n")

		// Find where the link footer starts: scan from end, stop at first non-empty non-link line.
		footerStart := -1
		for i, v := range slices.Backward(bodyLines) {
			if linkDefRE.MatchString(v) {
				footerStart = i
			} else if strings.TrimSpace(v) != "" {
				break
			}
		}

		if footerStart != -1 {
			footer = strings.Join(bodyLines[footerStart:], "\n")
			lastSection.body = strings.Join(bodyLines[:footerStart], "\n")
		}
	}

	return parsedChangelog{preamble: preamble, sections: sections, footer: footer}
}

// serialiseChangelog converts a parsedChangelog back to a string.
func serialiseChangelog(parsed parsedChangelog) string {
	var parts []string

	if strings.TrimSpace(parsed.preamble) != "" {
		parts = append(parts, parsed.preamble)
	}

	for _, section := range parsed.sections {
		parts = append(parts, section.header)
		if section.bodySet {
			parts = append(parts, section.body)
		}
	}

	if strings.TrimSpace(parsed.footer) != "" {
		parts = append(parts, parsed.footer)
	}

	return strings.Join(parts, "\n")
}

// normaliseSectionBody normalises new section body content:
//   - Strips leading blank lines
//   - Trims trailing whitespace/newlines
//   - Adds a leading "\n" and trailing "\n"
func normaliseSectionBody(content string) string {
	trimmed := strings.TrimRight(regexp.MustCompile(`^\n+`).ReplaceAllString(content, ""), " \t\n\r")
	return "\n" + trimmed + "\n"
}

// ---------------------------------------------------------------------------
// Rewrite operations
// ---------------------------------------------------------------------------

// rewriteUnreleased replaces the ## [Unreleased] section body.
func rewriteUnreleased(changelogContent, newBody string) (string, error) {
	parsed := parseChangelog(changelogContent)

	idx := -1
	for i, s := range parsed.sections {
		if s.header == headerUnreleased {
			idx = i
			break
		}
	}
	if idx == -1 {
		return "", fmt.Errorf("## [Unreleased] section not found in CHANGELOG.md")
	}

	parsed.sections[idx].body = normaliseSectionBody(newBody)
	parsed.sections[idx].bodySet = true

	return serialiseChangelog(parsed), nil
}

// rewriteRelease replaces or inserts the ## [x.y.z] - YYYY-MM-DD section.
// The ## [Unreleased] section is left unchanged.
func rewriteRelease(changelogContent, targetVersion, releaseDate, newBody string) string {
	parsed := parseChangelog(changelogContent)

	releaseHeader := fmt.Sprintf("## [%s] - %s", targetVersion, releaseDate)
	headerPrefix := fmt.Sprintf("## [%s]", targetVersion)

	// Find existing release section (match by version prefix, ignoring date)
	releaseIdx := -1
	for i, s := range parsed.sections {
		if strings.HasPrefix(s.header, headerPrefix) {
			releaseIdx = i
			break
		}
	}

	if releaseIdx == -1 {
		// Insert after ## [Unreleased]
		unreleasedIdx := -1
		for i, s := range parsed.sections {
			if s.header == headerUnreleased {
				unreleasedIdx = i
				break
			}
		}

		newSection := changelogSection{
			header:  releaseHeader,
			body:    normaliseSectionBody(newBody),
			bodySet: true,
		}

		if unreleasedIdx != -1 {
			// Insert after unreleased
			newSections := make([]changelogSection, 0, len(parsed.sections)+1)
			newSections = append(newSections, parsed.sections[:unreleasedIdx+1]...)
			newSections = append(newSections, newSection)
			newSections = append(newSections, parsed.sections[unreleasedIdx+1:]...)
			parsed.sections = newSections
		} else {
			// Prepend
			parsed.sections = append([]changelogSection{newSection}, parsed.sections...)
		}
	} else {
		parsed.sections[releaseIdx].header = releaseHeader
		parsed.sections[releaseIdx].body = normaliseSectionBody(newBody)
		parsed.sections[releaseIdx].bodySet = true
	}

	return serialiseChangelog(parsed)
}

// rewriteSectionResult is the JSON output of the rewrite operation.
type rewriteSectionResult struct {
	Success       bool   `json:"success"`
	ChangelogPath string `json:"changelogPath"`
	Section       string `json:"section"`
	Error         string `json:"error,omitempty"`
}

// rewriteSection is the unified entry point: rewrites the appropriate section based on mode.
func rewriteSection(changelogPath, mode, targetVersion, releaseDate, newSectionBody string) (rewriteSectionResult, error) {
	absPath, err := filepath.Abs(changelogPath)
	if err != nil {
		return rewriteSectionResult{}, err
	}

	changelogContent, err := os.ReadFile(absPath)
	if err != nil {
		return rewriteSectionResult{}, err
	}

	date := releaseDate
	if date == "" {
		date = time.Now().Format("2006-01-02")
	}

	var updated string
	var sectionHeader string

	if mode == "release" {
		if targetVersion == "" {
			return rewriteSectionResult{}, fmt.Errorf("targetVersion is required when mode == \"release\"")
		}
		updated = rewriteRelease(string(changelogContent), targetVersion, date, newSectionBody)
		sectionHeader = fmt.Sprintf("## [%s] - %s", targetVersion, date)
	} else {
		updated, err = rewriteUnreleased(string(changelogContent), newSectionBody)
		if err != nil {
			return rewriteSectionResult{}, err
		}
		sectionHeader = headerUnreleased
	}

	if err := os.WriteFile(absPath, []byte(updated), 0644); err != nil {
		return rewriteSectionResult{}, err
	}

	return rewriteSectionResult{Success: true, ChangelogPath: absPath, Section: sectionHeader}, nil
}

// ---------------------------------------------------------------------------
// CLI entry point
// ---------------------------------------------------------------------------

func runRewriteChangelogSection(args []string) {
	argMap := parseArgMap(args)

	mode := argMapGet(argMap, "mode", os.Getenv("MODE"), "unreleased")
	targetVersion := argMapGet(argMap, "target-version", os.Getenv("TARGET_VERSION"), "")
	changelogPath := argMapGet(argMap, "changelog", os.Getenv("CHANGELOG_PATH"), "CHANGELOG.md")
	sectionFile := argMapGet(argMap, "section-file", os.Getenv("SECTION_FILE"), "")

	newSectionBody := argMapGet(argMap, "section-content", os.Getenv("SECTION_CONTENT"), "")

	writeResult := func(r rewriteSectionResult) {
		out, _ := json.MarshalIndent(r, "", "  ")
		fmt.Println(string(out))
	}

	if newSectionBody == "" && sectionFile != "" {
		raw, err := os.ReadFile(sectionFile)
		if err != nil {
			r := rewriteSectionResult{
				Success:       false,
				ChangelogPath: changelogPath,
				Section:       "",
				Error:         fmt.Sprintf("Cannot read section file: %s", err.Error()),
			}
			writeResult(r)
			os.Exit(1)
		}
		newSectionBody = string(raw)
	}

	if newSectionBody == "" {
		r := rewriteSectionResult{
			Success:       false,
			ChangelogPath: changelogPath,
			Section:       "",
			Error:         "No section content provided (use --section-content or --section-file)",
		}
		writeResult(r)
		os.Exit(1)
	}

	result, err := rewriteSection(changelogPath, mode, targetVersion, "", newSectionBody)
	if err != nil {
		r := rewriteSectionResult{
			Success:       false,
			ChangelogPath: changelogPath,
			Section:       "",
			Error:         err.Error(),
		}
		writeResult(r)
		os.Exit(1)
	}

	writeResult(result)
}
